cpp/src/arrow/array/array_base.cc

```diff
@@ -165,7 +175,7 @@ struct ScalarFromArraySlotImpl {
                                 array_.length());
     }
 
-    if (array_.IsNull(index_)) {
+    if (array_.type()->id() != Type::RUN_END_ENCODED && array_.IsNull(index_)) {
```

Comment:
Hmm, I suppose this is unavoidable.

Reply:
what do you mean?

Reply:
I think the check is much cheaper than making `IsNull` virtually-dispatched.

Reply:
Right, it's just an observation...